### ISSUE TITLE
Negative lookahead caused java.util.NoSuchElementException

### ISSUE THREAD

**Issue by tomaszwojciechowski** [cucumber/cucumber#510]
On cucumber-java 3.0.2, I use cucumber expression:
` ("^(.+): In (.+) I type ((?!value of).+)")`
and this negative lookahead causes java.util.NoSuchElementException.

**Comment by Lenence** [comment]
Hello
The same thing here :
`A ((?!validated).+) on item (.+)`

I try with version 4.1.1 of cucumber-java and I have always the problem.

**Comment by mkq** [comment]
I guess the NoSuchElementException I'm getting is the same problem. In my case caused by an embedded flag in the ParameterType regex.

Here is a test case. The first step uses a ParameterType with a simple regex and passes. The second uses an equivalent regex with flag `(?x)` and fails.
#### pom.xml
```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>test</groupId>
  <artifactId>cucumber-parametertype-regex</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  <dependencies>
      <dependency>
          <groupId>io.cucumber</groupId>
          <artifactId>cucumber-java</artifactId>
          <version>4.2.3</version>
      </dependency>
      <dependency>
          <groupId>io.cucumber</groupId>
          <artifactId>cucumber-junit</artifactId>
          <version>4.2.3</version>
      </dependency>
  </dependencies>
</project>
```
#### src/test/resources/test/parametertype-regex.feature
```
Feature: custom ParameterTypes using regular expression with and without embedded flag

  Scenario: custom ParameterTypes using regular expression without embedded flag
    Then A: 4 is between -10 and 5

  Scenario: custom ParameterTypes using regular expression with embedded flag
    Then B: 4 is between -10 and 5
```
#### src/test/java/test/CustomParameterTypeSteps.java
```
package test;
import static org.junit.Assert.assertTrue;
import cucumber.api.java.en.Then;

public class CustomParameterTypeSteps {
        @Then("A: {int} is between {closedIntIntervalA}")
        public void isBetweenA(int value, ClosedIntInterval interval) {
                assertTrue(interval.contains(value));
        }

        @Then("B: {int} is between {closedIntIntervalB}")
        public void isBetweenB(int value, ClosedIntInterval interval) {
                assertTrue(interval.contains(value));
        }
}
```
#### src/test/java/test/TypeRegistryConfiguration.java
```
package test;
import cucumber.api.TypeRegistry;
import cucumber.api.TypeRegistryConfigurer;
import io.cucumber.cucumberexpressions.CaptureGroupTransformer;
import io.cucumber.cucumberexpressions.ParameterType;
import java.util.Locale;

public class TypeRegistryConfiguration implements TypeRegistryConfigurer {
        @Override
        public Locale locale() { return Locale.getDefault(); }

        @Override
        public void configureTypeRegistry(TypeRegistry typeRegistry) {
                CaptureGroupTransformer<ClosedIntInterval> ciiTransformer = groups -> ClosedIntInterval.of(Integer.parseInt(groups[0]),
                    Integer.parseInt(groups[1]));
                typeRegistry.defineParameterType(
                    new ParameterType<>("closedIntIntervalA", "(-?\\d+)\\s+and\\s+(-?\\d+)", ClosedIntInterval.class, ciiTransformer));
                typeRegistry.defineParameterType(
                    new ParameterType<>("closedIntIntervalB", "(?x) (-?\\d+) \\s+ and \\s+ (-?\\d+)", ClosedIntInterval.class, ciiTransformer));
        }
}
```
#### src/test/java/test/ClosedIntInterval.java
```
package test;

public class ClosedIntInterval {
        private final int start, end;

        private ClosedIntInterval(int start, int end) {
                this.start = start;
                this.end = end;
        }

        public static ClosedIntInterval of(int start, int end) {
                return new ClosedIntInterval(start, end);
        }

        public boolean contains(int value) {
                return start <= value && value <= end;
        }
}
```
#### src/test/java/test/CucumberTest.java
```
package test;
import cucumber.api.junit.Cucumber;
import org.junit.runner.RunWith;

@RunWith(Cucumber.class)
public class CucumberTest {}
```

**Comment by nzifnab** [comment]
I suspect this is related to https://github.com/cucumber/cucumber/issues/481

**Comment by mpkorstanje** [comment]
Yup. And was fixed by #593.
